src/main/java/com/conveyal/datatools/manager/controllers/api/LabelController.java

```diff
@@ -0,0 +1,217 @@
+                // Only return labels user has access to
+                .filter(label -> isProjectAdmin || !label.adminOnly)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * HTTP endpoint to create a new label
+     */
+    private static Label createLabel(Request req, Response res) throws IOException {
+        Auth0UserProfile userProfile = req.attribute("user");
+        Label newLabel = getPOJOFromRequestBody(req, Label.class);
+        validate(req, newLabel);
+        // User may not be allowed to create a new label
+        checkLabelPermissions(req, newLabel, Actions.CREATE);
+
+        try {
+            Persistence.labels.create(newLabel);
+            return newLabel;
+        } catch (Exception e) {
+            logMessageAndHalt(req, 500, "Unknown error encountered creating label.", e);
+            return null;
+        }
+    }
+
+    /**
+     * Check that updated or new label object is valid. This method should be called before a label is
+     * persisted to the database.
+     */
+    private static void validate(Request req, Label label) {
+        // Label is quite forgiving (sets defaults if null) and the boolean value is type checked,
+        // so there is little to validate.
+        if (StringUtils.isEmpty(label.name)) {
+            logMessageAndHalt(
+                    req,
+                    HttpStatus.BAD_REQUEST_400,
+                    "Request was invalid, the name field must not be empty."
+            );
+        }
+    }
+
+    /**
+     * Spark HTTP endpoint to update a label. Requires that the JSON body represent all
+     * fields the updated label should contain. See a similar note on FeedSourceController for more info
+     */
+    private static Label updateLabel(Request req, Response res) throws IOException {
+        String labelId = req.params("id");
+        Label formerLabel = requestLabelById(req, Actions.MANAGE);
+        Label updatedLabel = getPOJOFromRequestBody(req, Label.class);
+
+        // Some things shouldn't be updated
+        updatedLabel.projectId = formerLabel.projectId;
+        updatedLabel.id = formerLabel.id;
+
+        validate(req, updatedLabel);
+
+        Persistence.labels.replace(labelId, updatedLabel);
+
+        return updatedLabel;
+    }
+    /**
+     * Spark HTTP endpoint to delete a label
+     *
+     * FIXME: Should this just set a "deleted" flag instead of removing from the database entirely?
+     */
+    private static Label deleteLabel(Request req, Response res) {
+        Label label = requestLabelById(req, Actions.MANAGE);
+
+        // More specific error message if label doesn't exist
+        if (label == null) {
+            logMessageAndHalt(req, 403, "Label does not exist.");
+            return null;
+        }
+
+        try {
+            // Iterate over feed sources and remove any references to the label
+            Persistence.feedSources.getAll().forEach(feedSource -> {
```

Comment:
Please fetch just the feed sources according to the label's project ID from the database.

Reply:
Have refactored this in both fa4ba57 and then further in 9ac824a. I think it's more efficient, but would appreciate your double-check